pkg/gameserverallocations/controller.go

```diff
@@ -0,0 +1,394 @@
+// Runs threadiness number workers to process the rate limited queue
+// Probably only needs 1 worker, as its just deleting unallocated GameServerAllocations
+func (c *Controller) Run(workers int, stop <-chan struct{}) error {
+	err := crd.WaitForEstablishedCRD(c.crdGetter, "gameserverallocations."+stable.GroupName, c.logger)
+	if err != nil {
+		return err
+	}
+
+	err = c.counter.Run(stop)
+	if err != nil {
+		return err
+	}
+
+	c.stop = stop
+
+	c.logger.Info("Wait for cache sync")
+	if !cache.WaitForCacheSync(stop, c.gameServerAllocationSynced) {
+		return errors.New("failed to wait for caches to sync")
+	}
+
+	c.workerqueue.Run(workers, stop)
+	return nil
+}
+
+// creationMutationHandler will intercept when a GameServerAllocation is created, and allocate it a GameServer
+// assuming that one is available. If not, it will reject the AdmissionReview.
+func (c *Controller) creationMutationHandler(review admv1beta1.AdmissionReview) (admv1beta1.AdmissionReview, error) {
+	c.logger.WithField("review", review).Info("creationMutationHandler")
+	obj := review.Request.Object
+	gsa := &v1alpha1.GameServerAllocation{}
+
+	err := json.Unmarshal(obj.Raw, gsa)
+	if err != nil {
+		c.logger.WithError(err).Error("error unmarchaslling json")
+		return review, errors.Wrapf(err, "error unmarshalling original GameServerAllocation json: %s", obj.Raw)
+	}
+
+	gsa.ApplyDefaults()
+	gs, err := c.allocate(gsa)
+	if err != nil && err != ErrNoGameServerReady {
+		return review, err
+	}
+
+	if err == ErrNoGameServerReady {
+		gsa.Status.State = v1alpha1.GameServerAllocationUnAllocated
+	} else {
+		// When a GameServer is deleted, the GameServerAllocation should go with it
+		ref := metav1.NewControllerRef(gs, v1alpha1.SchemeGroupVersion.WithKind("GameServer"))
+		gsa.ObjectMeta.OwnerReferences = append(gsa.ObjectMeta.OwnerReferences, *ref)
+		gsa.Status.State = v1alpha1.GameServerAllocationAllocated
+		gsa.Status.GameServerName = gs.ObjectMeta.Name
+		gsa.Status.Ports = gs.Status.Ports
+		gsa.Status.Address = gs.Status.Address
+		gsa.Status.NodeName = gs.Status.NodeName
+	}
+
+	newFA, err := json.Marshal(gsa)
+	if err != nil {
+		c.logger.WithError(err).Error("error marshalling")
+		return review, errors.Wrapf(err, "error marshalling GameServerAllocation %s to json", gsa.ObjectMeta.Name)
+	}
+
+	patch, err := jsonpatch.CreatePatch(obj.Raw, newFA)
+	if err != nil {
+		c.logger.WithError(err).Error("error creating the patch")
+		return review, errors.Wrapf(err, "error creating patch for GameServerAllocation %s", gsa.ObjectMeta.Name)
+	}
+
+	json, err := json.Marshal(patch)
+	if err != nil {
+		c.logger.WithError(err).Error("error creating the json for the patch")
+		return review, errors.Wrapf(err, "error creating json for patch for GameServerAllocation %s", gs.ObjectMeta.Name)
+	}
+
+	c.logger.WithField("gsa", gsa.ObjectMeta.Name).WithField("patch", string(json)).Infof("patch created!")
+
+	pt := admv1beta1.PatchTypeJSONPatch
```

Comment:
I'm curious what this does ? Tells the webhook to patch the resource ?

Reply:
So when you return the `AdmissionReview.Response`, you can apply a `Patch, which states what changes you want to make to the object that you were originally passed - that's how you make changes via the mutationWebHook.

https://medium.com/ibm-cloud/diving-into-kubernetes-mutatingadmissionwebhook-6ef3c5695f74 has a good explanation.